Source/Diagnostics/ReducedDiags/ParticleHistogram.cpp

```diff
@@ -0,0 +1,170 @@
+
+#include "ParticleHistogram.H"
+#include "WarpX.H"
+#include "WarpXConst.H"
+#include "WarpXUtil.H"
+#include "AMReX_REAL.H"
+#include "AMReX_ParticleReduce.H"
+#include <limits> // numeric_limits<float>::min()
+
+using namespace amrex;
+
+// constructor
+ParticleHistogram::ParticleHistogram (std::string rd_name)
+: ReducedDiags{rd_name}
+{
+
+    // read parameters
+    ParmParse pp(rd_name);
+    pp.get("species",m_species_name);
+    pp.get("bin_number",m_bin_num);
+    pp.get("bin_max",   m_bin_max);
+    pp.get("bin_min",   m_bin_min);
+    pp.query("normalization",m_norm);
+    m_bin_size = (m_bin_max - m_bin_min) / m_bin_num;
+    std::string function_string = "";
+    Store_parserString(pp,"histogram_function(t,x,y,z,ux,uy,uz)",
+                       function_string);
+    m_parser.reset(new ParserWrapper<7>(
+        makeParser(function_string,{"t","x","y","z","ux","uy","uz"})));
+
+    // resize data array
+    m_data.resize(m_bin_num,0.0);
+
+    if (ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs;
+            ofs.open(m_path + m_rd_name + "." + m_extension,
+                std::ofstream::out | std::ofstream::app);
+            // write header row
+            ofs << "#";
+            ofs << "[1]step()";
+            ofs << m_sep;
+            ofs << "[2]time(s)";
+            for (int i = 0; i < m_bin_num; ++i)
+            {
+                ofs << m_sep;
+                ofs << "[" + std::to_string(3+i) + "]";
+                ofs << "bin" + std::to_string(1+i) + "()";
+            }
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+
+}
+// end constructor
+
+// function that computes the histogram
+void ParticleHistogram::ComputeDiags (int step)
+{
+
+    // Judge if the diags should be done
+    if ( (step+1) % m_freq != 0 ) { return; }
+
+    // get WarpX class object
+    auto & warpx = WarpX::GetInstance();
+
+    // get MultiParticleContainer class object
+    auto & mypc = warpx.GetPartContainer();
+
+    // get number of species (int)
+    auto nSpecies = mypc.nSpecies();
+
+    // get species names (std::vector<std::string>)
+    auto species_names = mypc.GetSpeciesNames();
+
+    // get time at level 0
+    auto t = warpx.gett_new(0);
+
+    // select species
+    int i_s;
+    for ( int i = 0; i < nSpecies; ++i )
+    {
+        if ( m_species_name == species_names[i] ) { i_s = i; }
+    }
+
+    // get WarpXParticleContainer class object
+    auto & myspc = mypc.GetParticleContainer(i_s);
+
+    using PType = typename WarpXParticleContainer::SuperParticleType;
+
+    // get parser
+    ParserWrapper<7> *fun_partparser = m_parser.get();
+
+    for ( int i = 0; i < m_bin_num; ++i )
+    {
+        // compute the histogram
+        m_data[i] = ReduceSum( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+        {
+            auto w  = p.rdata(PIdx::w);
```

Comment:
try to make all of those `auto const` please, to assist the compiler to optimize this :)

Reply:
Thanks for the suggestion, I will add `const` in the PR #727 .